include/deal.II/base/symmetric_tensor.h

```diff
@@ -2320,6 +2324,1003 @@ Number second_invariant (const SymmetricTensor<2,3,Number> &t)
+        }
+      else                      // This is the standard branch
+        {
+          norm = std::sqrt(1.0 / norm);
+          for (unsigned j=0; j < dim; j++)
+            Q[j][0] = Q[j][0] * norm;
+        }
+
+      // Calculate second eigenvector by the formula
+      //   v[1] = (A - w[1]).e1 x (A - w[1]).e2
+      Q[0][1]  = Q[0][1] + A[0][2]*w[1];
+      Q[1][1]  = Q[1][1] + A[1][2]*w[1];
+      Q[2][1]  = (A[0][0] - w[1]) * (A[1][1] - w[1]) - Q[2][1];
+      norm     = Q[0][1]*Q[0][1] + Q[1][1]*Q[1][1] + Q[2][1]*Q[2][1];
+      if (norm <= error)
+        {
+          return ql_implicit_shifts(A);
+        }
+      else
+        {
+          norm = std::sqrt(1.0 / norm);
+          for (unsigned int j=0; j < dim; j++)
+            Q[j][1] = Q[j][1] * norm;
+        }
+
+      // Calculate third eigenvector according to
+      //   v[2] = v[0] x v[1]
+      Q[0][2] = Q[1][0]*Q[2][1] - Q[2][0]*Q[1][1];
+      Q[1][2] = Q[2][0]*Q[0][1] - Q[0][0]*Q[2][1];
+      Q[2][2] = Q[0][0]*Q[1][1] - Q[1][0]*Q[0][1];
+
+      // Structure the data to be outputted
+      std::array<std::pair<Number, Tensor<1,dim,Number> >,dim> eig_vals_vecs;
+      for (unsigned int e=0; e<dim; ++e)
+        {
+          eig_vals_vecs[e].first = w[e];
+
+          // The column "e" of Q contains the non-normalized
+          // eigenvector associated with the eigenvalue "e"
+          for (unsigned int a=0; a<dim; ++a)
+            {
+              eig_vals_vecs[e].second[a] = Q[a][e];
+            }
+
+          // Normalize
+          Assert(eig_vals_vecs[e].second.norm() != 0.0, ExcDivideByZero());
+          eig_vals_vecs[e].second /= eig_vals_vecs[e].second.norm();
+        }
+      return eig_vals_vecs;
+    }
+
+
+    template<int dim, typename Number>
+    struct SortEigenValuesVectors
+    {
+      typedef std::pair<Number, Tensor<1,dim,Number> > EigValsVecs;
+      bool operator() (const EigValsVecs &lhs,
+                       const EigValsVecs &rhs)
+      {
+        return lhs.first > rhs.first;
+      }
+    };
+
+  }
+} // namespace internal
+
+
+
+/**
+ * An enumeration for the algorithm to be employed when performing
+ * the computation of normalized eigenvectors and their corresponding
+ * eigenvalues.
+ *
+ * The specialized algorithms utilized in computing the eigenvectors are
+ * presented in
+ * @code{.bib}
+ * @Article{Kopp2008,
+ *   title        = {Efficient numerical diagonalization of hermitian 3x3 matrices},
+ *   author       = {Kopp, J.},
+ *   journal      = {International Journal of Modern Physics C},
+ *   year         = {2008},
+ *   volume       = {19},
+ *   number       = {3},
+ *   pages        = {523--548},
+ *   doi          = {10.1142/S0129183108012303},
+ *   eprinttype   = {arXiv},
+ *   eprint       = {physics/0610206v3},
+ *   eprintclass  = {physics.comp-ph},
+ *   url          = {https://www.mpi-hd.mpg.de/personalhomes/globes/3x3/index.html}
+ * }
+ * @endcode
+ */
+enum eigenvector_method
```

Comment:
It may have been a historical mistake to put `invert`, `trace`, etc into the global namespace. But I see your point that for conceptual clarity it makes sense to treat like functions alike.

Good point also about the global enumerations.